### PR TITLE
Upgraded to MongoDb java driver 4.0.3

### PR DESCRIPTION
**Pull request by bjornharvold**
Here's a PR with support for the latest MongoDb Java driver 4.0.3. I noted @deprecated or added a TODO wherever necessary.

Note, this is not backwards compatible.

Please have a look. If all good, would love to get a release of this so I can upgrade to Spring Boot 2.3.0 :-)

Cheers,
Bjorn

### CODE REVIEW

File: mongo/src/main/java/org/axonframework/extensions/mongo/serialization/DBObjectToStringContentTypeConverter.java [hunk 1]
```diff
@@ -41,6 +41,7 @@
 
     @Override
     public String convert(DBObject original) {
-        return JSON.serialize(original);
+        // TODO assuming this is BasicDbObject always
```
Comment (smcvb) [inline mongo/src/main/java/org/axonframework/extensions/mongo/serialization/DBObjectToStringContentTypeConverter.java:44]: [This](https://mongodb.github.io/mongo-java-driver/3.6/javadoc/index.html?com/mongodb/util/JSON.html) API page states the `JSON` utility was replaced by the `BasicDBObject` toJson/parse methods. Hence your assumption is right and as such you can drop this line of commentary.

File: mongo/src/test/java/org/axonframework/extensions/mongo/eventsourcing/eventstore/MongoOptionsFactoryTest.java [hunk 2]
```diff
@@ -16,9 +16,11 @@
...
 
-import com.mongodb.MongoClientOptions;
+import com.mongodb.MongoClientSettings;
 import org.junit.jupiter.api.*;
 
+import java.util.concurrent.TimeUnit;
+
 import static org.junit.jupiter.api.Assertions.*;
 
 /**
```
Comment (smcvb) [inline mongo/src/test/java/org/axonframework/extensions/mongo/eventsourcing/eventstore/MongoOptionsFactoryTest.java:26]: We should rename this class to `MongoSettingsFactoryTest` to be inline with the object it tests.

File: mongo/src/test/java/org/axonframework/extensions/mongo/eventsourcing/eventstore/MongoOptionsFactoryTest.java [hunk 3]
```diff
@@ -37,15 +39,17 @@ void setUp() {
...
-        assertEquals(defaults.getMaxWaitTime(), options.getMaxWaitTime());
-        assertEquals(defaults.getSocketTimeout(), options.getSocketTimeout());
-        assertEquals(defaults.getConnectionsPerHost(), options.getConnectionsPerHost());
-        assertEquals(defaults.getConnectTimeout(), options.getConnectTimeout());
-        assertEquals(defaults.getThreadsAllowedToBlockForConnectionMultiplier(),
-                     options.getThreadsAllowedToBlockForConnectionMultiplier());
+        assertEquals(defaults.getConnectionPoolSettings().getMaxWaitTime(TimeUnit.MILLISECONDS), options.getConnectionPoolSettings().getMaxWaitTime(TimeUnit.MILLISECONDS));
+        assertEquals(defaults.getSocketSettings().getConnectTimeout(TimeUnit.MILLISECONDS), options.getSocketSettings().getConnectTimeout(TimeUnit.MILLISECONDS));
+        assertEquals(defaults.getConnectionPoolSettings().getMaxSize(), options.getConnectionPoolSettings().getMaxSize());
+        // TODO connect time and socket timeout is identical
```
Comment (smcvb) [inline mongo/src/test/java/org/axonframework/extensions/mongo/eventsourcing/eventstore/MongoOptionsFactoryTest.java:48]: Are you certain these are identical? If so, do you per chance have any material on this to reference for me?

File: mongo/src/test/java/org/axonframework/extensions/mongo/eventsourcing/eventstore/MongoOptionsFactoryTest.java [hunk 4]
```diff
@@ -37,15 +39,17 @@ void setUp() {
...
+//        assertEquals(defaults.getConnectTimeout(), options.getConnectTimeout());
+        // TODO deprecated
```
Comment (smcvb) [inline mongo/src/test/java/org/axonframework/extensions/mongo/eventsourcing/eventstore/MongoOptionsFactoryTest.java:50]: We can drop this assertion as the `getThreadsAllowedToBlockForConnectionMultiplier` logic will be removed too.

File: mongo/src/test/java/org/axonframework/extensions/mongo/eventsourcing/eventstore/MongoOptionsFactoryTest.java [hunk 5]
```diff
@@ -54,13 +58,16 @@ void testCreateMongoOptions_customSet() {
         factory.setConnectionTimeout(11);
         factory.setMaxWaitTime(3);
         factory.setSocketTimeOut(23);
-        factory.setThreadsAllowedToBlockForConnectionMultiplier(31);
+//        factory.setThreadsAllowedToBlockForConnectionMultiplier(31);
```
Comment (smcvb) [inline mongo/src/test/java/org/axonframework/extensions/mongo/eventsourcing/eventstore/MongoOptionsFactoryTest.java:61]: We can drop this assertion as the `setThreadsAllowedToBlockForConnectionMultiplier` logic will be removed too.

File: mongo/src/test/java/org/axonframework/extensions/mongo/eventsourcing/eventstore/MongoOptionsFactoryTest.java [hunk 6]
```diff
@@ -54,13 +58,16 @@ void testCreateMongoOptions_customSet() {
...
-        MongoClientOptions options = factory.createMongoOptions();
-        assertEquals(3, options.getMaxWaitTime());
-        assertEquals(23, options.getSocketTimeout());
-        assertEquals(9, options.getConnectionsPerHost());
-        assertEquals(11, options.getConnectTimeout());
-        assertEquals(31, options.getThreadsAllowedToBlockForConnectionMultiplier());
+        // TODO connect time and socket timeout is identical
+//        assertEquals(11, options.getConnectTimeout());
+        // TODO deprecated
+//        assertEquals(31, options.getThreadsAllowedToBlockForConnectionMultiplier());
```
Comment (smcvb) [inline mongo/src/test/java/org/axonframework/extensions/mongo/eventsourcing/eventstore/MongoOptionsFactoryTest.java:71]: We can drop this assertion as the `getThreadsAllowedToBlockForConnectionMultiplier` logic will be removed too.

File: pom.xml [hunk 7]
```diff
@@ -405,6 +405,13 @@
     </build>
 
     <repositories>
+        <!-- Maven repository -->
+<!--        <repository>-->
```
Comment (smcvb) [inline pom.xml:409]: I am guessing this wasn't intended to be part of the pull request? Of so, please remove it in a subsequent commit.

File: mongo/src/test/java/org/axonframework/extensions/mongo/eventsourcing/eventstore/MongoOptionsFactoryTest.java (hunk 3 above)
Comment (bjornharvold) [reply mongo/src/test/java/org/axonframework/extensions/mongo/eventsourcing/eventstore/MongoOptionsFactoryTest.java]: There was no document that described the changes made and how to use the new methods in the MongoDb docs so I had to go comparing the old docs with the new docs.

http://mongodb.github.io/mongo-java-driver/4.0/apidocs/mongodb-driver-core/com/mongodb/connection/ConnectionPoolSettings.html
http://mongodb.github.io/mongo-java-driver/4.0/apidocs/mongodb-driver-core/com/mongodb/connection/SocketSettings.html
http://mongodb.github.io/mongo-java-driver/3.12/javadoc/com/mongodb/MongoClientOptions.html

**Comment by bjornharvold** [comment]
The PR has been updated with the comments on this review

File: mongo/src/main/java/org/axonframework/extensions/mongo/eventsourcing/eventstore/MongoFactory.java [hunk 8]
```diff
@@ -80,36 +51,8 @@ public void setMongoAddresses(List<ServerAddress> mongoAddresses) {
      *
      * @param mongoOptions MongoOptions to overrule the default
      */
-    public void setMongoOptions(MongoClientOptions mongoOptions) {
+    public void setMongoClientSettings(MongoClientSettings mongoOptions) {
```
Comment (smcvb) [inline mongo/src/main/java/org/axonframework/extensions/mongo/eventsourcing/eventstore/MongoFactory.java:54]:
```suggestion
    public void setMongoClientSettings(MongoClientSettings mongoClientSettings) {
```

File: mongo/src/main/java/org/axonframework/extensions/mongo/eventsourcing/eventstore/MongoSettingsFactory.java [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+    public MongoSettingsFactory() {
+        defaults = MongoClientSettings.builder().build();
+    }
+
+    /**
+     * Uses the configured parameters to create a MongoOptions instance.
+     *
+     * @return MongoOptions instance based on the configured properties
+     */
+    public MongoClientSettings createMongoOptions() {
```
Comment (smcvb) [inline mongo/src/main/java/org/axonframework/extensions/mongo/eventsourcing/eventstore/MongoSettingsFactory.java:68]:
```suggestion
    public MongoClientSettings createMongoSettings() {
```

File: mongo/src/main/java/org/axonframework/extensions/mongo/eventsourcing/eventstore/MongoSettingsFactory.java [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+     * Set the multiplier for the amount of threads to block in relation to the maximum amount of connections.
+     *
+     * @param threadsAllowedToBlockForConnectionMultiplier
+     *            Number representing the multiplier of the amount of threads to block in relation to the connections
+     *            that are allowed.
+     * @deprecated in the next major release, wait queue size limitations will be removed
+     */
+//    public void setThreadsAllowedToBlockForConnectionMultiplier(int threadsAllowedToBlockForConnectionMultiplier) {
+//        this.threadsAllowedToBlockForConnectionMultiplier = threadsAllowedToBlockForConnectionMultiplier;
+//    }
```
Comment (smcvb) [inline mongo/src/main/java/org/axonframework/extensions/mongo/eventsourcing/eventstore/MongoSettingsFactory.java:182]: Let's remove all these lines about the "allowed threads to block" setting.

File: mongo/src/test/java/org/axonframework/extensions/mongo/MongoTestContext.java [hunk 11]
```diff
@@ -31,15 +31,15 @@ public MongoClient mongoClient(MongoFactory mongoFactory) {
     }
 
     @Bean
-    public MongoFactory mongoFactoryBean(MongoOptionsFactory mongoOptionsFactory) {
+    public MongoFactory mongoFactoryBean(MongoSettingsFactory mongoOptionsFactory) {
```
Comment (smcvb) [inline mongo/src/test/java/org/axonframework/extensions/mongo/MongoTestContext.java:34]:
```suggestion
    public MongoFactory mongoFactoryBean(MongoSettingsFactory mongoSettingsFactory) {
```

File: mongo/src/test/java/org/axonframework/extensions/mongo/MongoTestContext.java [hunk 12]
```diff
@@ -31,15 +31,15 @@ public MongoClient mongoClient(MongoFactory mongoFactory) {
...
         MongoFactory mongoFactory = new MongoFactory();
-        mongoFactory.setMongoOptions(mongoOptionsFactory.createMongoOptions());
+        mongoFactory.setMongoClientSettings(mongoOptionsFactory.createMongoOptions());
         return mongoFactory;
     }
 
     @Bean
-    public MongoOptionsFactory mongoOptionsFactory() {
-        MongoOptionsFactory mongoOptionsFactory = new MongoOptionsFactory();
+    public MongoSettingsFactory mongoOptionsFactory() {
```
Comment (smcvb) [inline mongo/src/test/java/org/axonframework/extensions/mongo/MongoTestContext.java:41]:
```suggestion
    public MongoSettingsFactory mongoSettingsFactory() {
```

File: mongo/src/test/java/org/axonframework/extensions/mongo/MongoTestContext.java [hunk 13]
```diff
@@ -31,15 +31,15 @@ public MongoClient mongoClient(MongoFactory mongoFactory) {
...
+        MongoSettingsFactory mongoOptionsFactory = new MongoSettingsFactory();
         mongoOptionsFactory.setConnectionsPerHost(100);
         return mongoOptionsFactory;
```
Comment (smcvb) [inline mongo/src/test/java/org/axonframework/extensions/mongo/MongoTestContext.java:44]:
```suggestion
        MongoSettingsFactory mongoSettingsFactory = new MongoSettingsFactory();
        mongoSettingsFactory.setConnectionsPerHost(100);
        return mongoSettingsFactory;
```

File: mongo/src/test/java/org/axonframework/extensions/mongo/eventsourcing/eventstore/MongoEventStorageEngineTest_DBObjectSerialization.java [hunk 14]
```diff
@@ -159,17 +159,18 @@ public MongoClient mongoClient(MongoFactory mongoFactory) {
         }
 
         @Bean
-        public MongoFactory mongoFactoryBean(MongoOptionsFactory mongoOptionsFactory) {
+        public MongoFactory mongoFactoryBean(MongoSettingsFactory mongoOptionsFactory) {
```
Comment (smcvb) [inline mongo/src/test/java/org/axonframework/extensions/mongo/eventsourcing/eventstore/MongoEventStorageEngineTest_DBObjectSerialization.java:162]:
```suggestion
        public MongoFactory mongoFactoryBean(MongoSettingsFactory mongoSettingsFactory) {
```

File: mongo/src/test/java/org/axonframework/extensions/mongo/eventsourcing/eventstore/MongoEventStorageEngineTest_DBObjectSerialization.java [hunk 15]
```diff
@@ -159,17 +159,18 @@ public MongoClient mongoClient(MongoFactory mongoFactory) {
...
         }
 
         @Bean
-        public MongoOptionsFactory mongoOptionsFactory() {
-            MongoOptionsFactory mongoOptionsFactory = new MongoOptionsFactory();
+        public MongoSettingsFactory mongoOptionsFactory() {
+            MongoSettingsFactory mongoOptionsFactory = new MongoSettingsFactory();
             mongoOptionsFactory.setConnectionsPerHost(100);
+            mongoOptionsFactory.setWriteConcern(WriteConcern.JOURNALED);
             return mongoOptionsFactory;
```
Comment (smcvb) [inline mongo/src/test/java/org/axonframework/extensions/mongo/eventsourcing/eventstore/MongoEventStorageEngineTest_DBObjectSerialization.java:174]:
```suggestion
        public MongoSettingsFactory mongoSettingsFactory() {
            MongoSettingsFactory mongoSettingsFactory = new MongoSettingsFactory();
            mongoSettingsFactory.setConnectionsPerHost(100);
            mongoSettingsFactory.setWriteConcern(WriteConcern.JOURNALED);
            return mongoSettingsFactory;
```

File: mongo/src/test/java/org/axonframework/extensions/mongo/eventsourcing/eventstore/MongoEventStorageEngineTest_DBObjectSerialization.java [hunk 16]
```diff
@@ -159,17 +159,18 @@ public MongoClient mongoClient(MongoFactory mongoFactory) {
         }
 
         @Bean
-        public MongoFactory mongoFactoryBean(MongoOptionsFactory mongoOptionsFactory) {
+        public MongoFactory mongoFactoryBean(MongoSettingsFactory mongoOptionsFactory) {
             MongoFactory mongoFactory = new MongoFactory();
-            mongoFactory.setMongoOptions(mongoOptionsFactory.createMongoOptions());
-            mongoFactory.setWriteConcern(WriteConcern.JOURNALED);
+            mongoFactory.setMongoClientSettings(mongoOptionsFactory.createMongoOptions());
```
Comment (smcvb) [inline mongo/src/test/java/org/axonframework/extensions/mongo/eventsourcing/eventstore/MongoEventStorageEngineTest_DBObjectSerialization.java:164]:
```suggestion
            mongoFactory.setMongoClientSettings(mongoSettingsFactory.createMongoOptions());
```

File: mongo/src/test/java/org/axonframework/extensions/mongo/eventsourcing/eventstore/MongoEventStorageEngineTest_DocPerCommit.java [hunk 17]
```diff
@@ -276,15 +276,15 @@ public MongoClient mongoClient(MongoFactory mongoFactory) {
         }
 
         @Bean
-        public MongoFactory mongoFactoryBean(MongoOptionsFactory mongoOptionsFactory) {
+        public MongoFactory mongoFactoryBean(MongoSettingsFactory mongoOptionsFactory) {
```
Comment (smcvb) [inline mongo/src/test/java/org/axonframework/extensions/mongo/eventsourcing/eventstore/MongoEventStorageEngineTest_DocPerCommit.java:279]:
```suggestion
        public MongoFactory mongoFactoryBean(MongoSettingsFactory mongoSettingsFactory) {
```

File: mongo/src/test/java/org/axonframework/extensions/mongo/eventsourcing/eventstore/MongoEventStorageEngineTest_DocPerCommit.java [hunk 18]
```diff
@@ -276,15 +276,15 @@ public MongoClient mongoClient(MongoFactory mongoFactory) {
...
             MongoFactory mongoFactory = new MongoFactory();
-            mongoFactory.setMongoOptions(mongoOptionsFactory.createMongoOptions());
+            mongoFactory.setMongoClientSettings(mongoOptionsFactory.createMongoOptions());
```
Comment (smcvb) [inline mongo/src/test/java/org/axonframework/extensions/mongo/eventsourcing/eventstore/MongoEventStorageEngineTest_DocPerCommit.java:281]:
```suggestion
            mongoFactory.setMongoClientSettings(mongoSettingsFactory.createMongoOptions());
```

File: mongo/src/test/java/org/axonframework/extensions/mongo/eventsourcing/eventstore/MongoEventStorageEngineTest_DocPerCommit.java [hunk 19]
```diff
@@ -276,15 +276,15 @@ public MongoClient mongoClient(MongoFactory mongoFactory) {
...
             return mongoFactory;
         }
 
         @Bean
-        public MongoOptionsFactory mongoOptionsFactory() {
-            MongoOptionsFactory mongoOptionsFactory = new MongoOptionsFactory();
+        public MongoSettingsFactory mongoOptionsFactory() {
+            MongoSettingsFactory mongoOptionsFactory = new MongoSettingsFactory();
             mongoOptionsFactory.setConnectionsPerHost(100);
             return mongoOptionsFactory;
```
Comment (smcvb) [inline mongo/src/test/java/org/axonframework/extensions/mongo/eventsourcing/eventstore/MongoEventStorageEngineTest_DocPerCommit.java:289]:
```suggestion
        public MongoSettingsFactory mongoSettingsFactory() {
            MongoSettingsFactory mongoSettingsFactory = new MongoSettingsFactory();
            mongoSettingsFactory.setConnectionsPerHost(100);
            return mongoSettingsFactory;
```

**Comment by bjornharvold** [comment]
Code committed

File: mongo/src/main/java/org/axonframework/extensions/mongo/eventsourcing/eventstore/MongoOptionsFactory.java [hunk 20]
```diff
@@ -33,41 +43,44 @@
...
-                .connectionsPerHost(getConnectionsPerHost())
-                .connectTimeout(getConnectionTimeout())
-                .maxWaitTime(getMaxWaitTime())
-                .threadsAllowedToBlockForConnectionMultiplier(getThreadsAllowedToBlockForConnectionMultiplier())
-                .socketTimeout(getSocketTimeOut()).build();
+    public MongoClientSettings createMongoOptions() {
+        MongoClientSettings settings = MongoClientSettings.builder()
+                .applyToConnectionPoolSettings(builder -> builder.maxWaitTime(getMaxWaitTime(), TimeUnit.MILLISECONDS).maxSize(getConnectionsPerHost()))
+                .applyToClusterSettings(builder -> builder.hosts(mongoAddresses))
+                .applyToSocketSettings(builder -> builder.connectTimeout(getSocketTimeOut(), TimeUnit.MILLISECONDS))
```
[the remark this answers is not in the document]
Comment (smcvb) [reply mongo/src/main/java/org/axonframework/extensions/mongo/eventsourcing/eventstore/MongoOptionsFactory.java]: What's your thought about this @bjornharvold? 
Maybe we should change the existing `getSocketTimeOut` to specify it's for the socket's `connectTimeOut`?

At the least, right now it's rather ambiguous whether the `MongoOptionsFactory`'s socket time out is used for:
1. Both read and socket time out fields in the `SocketSettings.Builder`
2. The read time out field in the `SocketSettings.Builder`
3. The connect time out field in the `SocketSettings.Builder`

Hence we should be specific which of the three is referenced, both through method/field naming and javadoc.

File: mongo/src/main/java/org/axonframework/extensions/mongo/eventsourcing/eventstore/MongoOptionsFactory.java [hunk 21]
```diff
@@ -147,21 +160,66 @@ public void setSocketTimeOut(int socketTimeOut) {
...
+//                : defaults.getThreadsAllowedToBlockForConnectionMultiplier();
+//    }
 
     /**
      * Set the multiplier for the amount of threads to block in relation to the maximum amount of connections.
      *
      * @param threadsAllowedToBlockForConnectionMultiplier
      *            Number representing the multiplier of the amount of threads to block in relation to the connections
      *            that are allowed.
+     * @deprecated in the next major release, wait queue size limitations will be removed
```
[the remark this answers is not in the document]
Comment (smcvb) [reply mongo/src/main/java/org/axonframework/extensions/mongo/eventsourcing/eventstore/MongoOptionsFactory.java]: I'd still like this removal to be done to be honest :-)

File: mongo/src/main/java/org/axonframework/extensions/mongo/eventsourcing/eventstore/MongoSettingsFactory.java [hunk 22]
```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright (c) 2010-2016. Axon Framework
```
Comment (smcvb) [inline mongo/src/main/java/org/axonframework/extensions/mongo/eventsourcing/eventstore/MongoSettingsFactory.java:2]: Nit: The copyright notice should transition the period between 2010 and 2020.

File: mongo/src/main/java/org/axonframework/extensions/mongo/eventsourcing/eventstore/MongoSettingsFactory.java [hunk 23]
```diff
@@ -0,0 +1,202 @@
...
+
+    private static final Logger logger = LoggerFactory.getLogger(MongoSettingsFactory.class);
+
+    private final MongoClientSettings defaults;
+    private List<ServerAddress> mongoAddresses = Collections.emptyList();
+    private WriteConcern writeConcern;
+    private int connectionsPerHost;
+    private int connectionTimeout;
+    private long maxWaitTime;
+    private int socketTimeOut;
```
Comment (smcvb) [inline mongo/src/main/java/org/axonframework/extensions/mongo/eventsourcing/eventstore/MongoSettingsFactory.java:54]: I still think we need to do something about having both the `connectionTimeout` and the `socketTimeOut`. 
As it currently stands, you're only using the `socketTimeOut` in the `MongoClientSettings`.

On top of that, we only set the `SocketSettings.connectTimeout` field, whilst previously we set a connection timeout _and_ socket timeout field. I do agree that Mongo's documentation on the matter is rather vague, so it'll be hard to deduce (without accessing their old implementation) what the original intent was of both fields.

As such I'd like to suggest the following:

1. Rename `connectionTimeout` to `socketConnectTimeout`. Adjust getter/setter and javadoc accordingly.
2. Rename `socketTimeOut` to `socketReadTimeout`. Adjust getter/setter and javadoc accordingly.
3. Use the `socketConnectTimeout` for `SocketBuilder.connectTimeout` method in the `createMongoClientSettings` function.
4. Use the `socketReadTimeout` for `SocketBuilder.readTimeout` method in the `createMongoClientSettings` function.
5. Adjust the debug logging in the `createMongoClientSettings` function to correctly reference all this information.

What's your thought about this @bjornharvold? Does this freedom of configuration make sense for your use case?

At the very least I hope to make things clear from this `MongoSettingsFactory` from here onward.

File: mongo/src/main/java/org/axonframework/extensions/mongo/serialization/DBObjectToStringContentTypeConverter.java (hunk 1 above)
Comment (smcvb) [reply mongo/src/main/java/org/axonframework/extensions/mongo/serialization/DBObjectToStringContentTypeConverter.java:44]: This TODO can still be removed.

File: mongo/src/main/java/org/axonframework/extensions/mongo/eventsourcing/eventstore/MongoSettingsFactory.java (hunk 23 above)
Comment (bjornharvold) [reply mongo/src/main/java/org/axonframework/extensions/mongo/eventsourcing/eventstore/MongoSettingsFactory.java]: This works.

**Comment by bjornharvold** [comment]
Update to PR review committed. Tests passing.

**Comment by smcvb** [comment]
Thanks for all the effort on this @bjornharvold!

**Pull request merged**